packages/astro/src/core/build/plugins/plugin-prerender.ts

```diff
@@ -1,87 +1,111 @@
+			// after we finish prerendering.
+			const nonPrerenderOnlyChunks = getNonPrerenderOnlyChunks(bundle, internals);
+			internals.prerenderOnlyChunks = Object.values(bundle).filter((chunk) => {
+				return chunk.type === 'chunk' && !nonPrerenderOnlyChunks.has(chunk);
+			}) as Rollup.OutputChunk[];
 		},
 	};
 }
 
+function getNonPrerenderOnlyChunks(bundle: Rollup.OutputBundle, internals: BuildInternals) {
+	const chunks = Object.values(bundle);
+
+	const prerenderOnlyEntryChunks = new Set<Rollup.OutputChunk>();
+	const nonPrerenderOnlyEntryChunks = new Set<Rollup.OutputChunk>();
+	for (const chunk of chunks) {
+		if (chunk.type === 'chunk' && (chunk.isEntry || chunk.isDynamicEntry)) {
+			// See if this entry chunk is prerendered, if so, skip it
+			if (chunk.facadeModuleId?.startsWith(ASTRO_PAGE_RESOLVED_MODULE_ID)) {
+				const pageDatas = getPagesFromVirtualModulePageName(
+					internals,
+					ASTRO_PAGE_RESOLVED_MODULE_ID,
+					chunk.facadeModuleId
+				);
+				const prerender = pageDatas.every((pageData) => pageData.route.prerender);
+				if (prerender) {
+					prerenderOnlyEntryChunks.add(chunk);
+					continue;
+				}
+			}
+			// Ideally we should record entries when `functionPerRoute` is enabled, but this breaks some tests
+			// that expect the entrypoint to still exist even if it should be unused.
+			// TODO: Revisit this so we can delete additional unused chunks
+			// else if (chunk.facadeModuleId?.startsWith(RESOLVED_SPLIT_MODULE_ID)) {
+			// 	const pageDatas = getPagesFromVirtualModulePageName(
+			// 		internals,
+			// 		RESOLVED_SPLIT_MODULE_ID,
+			// 		chunk.facadeModuleId
+			// 	);
+			// 	const prerender = pageDatas.every((pageData) => pageData.route.prerender);
+			// 	if (prerender) {
+			// 		prerenderOnlyEntryChunks.add(chunk);
+			// 		continue;
+			// 	}
+			// }
+
+			nonPrerenderOnlyEntryChunks.add(chunk);
+		}
+	}
+
+	// From the `nonPrerenderedEntryChunks`, we crawl all the imports/dynamicImports to find all
+	// other chunks that are use by the non-prerendered runtime
+	const nonPrerenderOnlyChunks = new Set();
+	for (const entryChunk of nonPrerenderOnlyEntryChunks) {
+		crawlChunk(entryChunk);
+	}
+
+	function crawlChunk(chunk: Rollup.OutputChunk) {
+		if (nonPrerenderOnlyChunks.has(chunk)) return;
+		nonPrerenderOnlyChunks.add(chunk);
+		for (const importFileName of chunk.imports) {
+			const importChunk = bundle[importFileName];
+			if (importChunk?.type === 'chunk') {
+				crawlChunk(importChunk);
+			}
+		}
+		for (const dynamicImportFileName of chunk.dynamicImports) {
+			const dynamicImportChunk = bundle[dynamicImportFileName];
+			// The main server entry (entry.mjs) may import a prerender-only entry chunk, we skip in this case
+			// to prevent incorrectly marking it as non-prerendered.
+			if (
+				dynamicImportChunk?.type === 'chunk' &&
+				!prerenderOnlyEntryChunks.has(dynamicImportChunk)
+			) {
+				crawlChunk(dynamicImportChunk);
+			}
+		}
```

Comment:
Can we avoid a recursion and use a loop to prevent memory pressure in crawling all these chunks?

Reply:
I'm not sure it's possible to use a loop here. The `crawlChunk` function uses `nonPrerenderOnlyChunks` to guard against processing already seen chunks though.

Reply:
Why not? Since we don't depend on the order of the nodes we encountered (it seems) we can have an array where we push the children chunks as we go. With a do/while, it should be achievable.

Why don't we use the guard **before** starting a new recursive crawling?

Reply:
I see it now and committed a refactor using set iterators. It's shorter now too. Let me know what you think.